docs/spec/BeakerSpec.tex

```diff
@@ -0,0 +1,994 @@
+
+As shown below, the maximum number of procedures is $2^{24} - 1 = 16,777,215$,
+therefore the maxium value of the key index is $16,777,215$, therefore the key
+index lies in the range $\left[1,1677215\right]$.
+
+For this reason the key index is specified as 24 bytes.
+
+\subsection{Procedure List}\label{procedure-table}
+The procedure list is simply an array of storage values. The first value is the
+length of this list, and the subsequent values are the procedure keys of the
+list.
+
+When a procedure is added to the kernel:
+
+\begin{enumerate}
+  \item The procedure data is added to the procedure heap (see
+  Section~\ref{procedure-heap}).
+  \item The procedure key is appended to the end of the array.
+  \item The length value (the first value) is incremented by one.
+\end{enumerate}
+
+When a procedure is deleted from the kernel:
+
+\begin{enumerate}
+  \item If the procedure key is the same as the Entry Procedure Key, abort and
+  throw an error.
+  \item If the procedure key does not exist in the list, abort and throw an
+  error.
+  \item The length value is decremented by one.
+  \item If the procedure being deleted is not the last in the list, the last in
+  the list is copied to overwrite the key being deleted. This also accounts for
+  the case of an empty list.
+\end{enumerate}
+
+It is important to note that none of these steps consider deletion (zeroing) of
+data. This is optional for efficiency and can be performed any time after the
+length value has been decremented.
+
+The procedure table is stored under the prefix \texttt{0x00 00 00 00 01}. The
+maximum number of keys held under this configuration is $2^{24} - 1 =
+16,777,215$ (1 is subtracted to account for the length value occupying a single
+space). This is one less than the total number of procedures that can be held by
+the kernel (which is $2^{24} = 16,777,216$).
+
+\begin{verbatim}
+  0x00 00 00 00 01 + 24 more bytes + 3 zero bytes
+\end{verbatim}
+
+\begin{table}[H]
+  \caption{Procedure table.}
+  \centering{}%
+  \begin{tabular}{l|p{0.4\textwidth}}
+    \hline
+    Storage Location & Description\tabularnewline
+    \hline
+    \hline
+    \texttt{0x00 00 00 00 01}     & length/number of procedures
+      \tabularnewline
+    \texttt{0x00 00 00 00 01} + keyIndex & procedure key/name: 24 bytes
+      \tabularnewline
+    ~~~$\vdots$ & capabilities in series \tabularnewline
+    \texttt{0x00 00 00 00 01} + n & procedure key/name for procedure $n$
+      \tabularnewline
+    \hline
+  \end{tabular}
+\end{table}
+
+\subsection{Procedure Heap}\label{procedure-heap}
+
+Procedure data is stored using the key \texttt{0x00 00 00 00 00}. This is
+combined with the procedure key to produce the following:
+
+\begin{verbatim}
+  0x00 00 00 00 00 + 24 bytes procedure key + 3 bytes data offset
+\end{verbatim}
+
+This leaves 3 bytes at the end of the storage location key. If the first of
+these bytes is 0x00, the storage key refers to various metadata of the
```

Comment:
For procedure metadata let's add a Procedure List Index for faster lookup.

| Data Offset | | |  | Description              |
|----|----|----|----|----------------------|
| 0x | 00 | 00 | 00 | Address              |
| 0x | 00 | 00 | 01 | Procedure List Index |

Reply:
See the comment below re: Key Index / Procedure Index.